### ISSUE TITLE
Question on how the offsets used for metadata

### ISSUE THREAD

**Issue by muraliran** [NetSys/bess#679]
I am looking at set_metadata as example on using snbuf metadata segment.

In method:
static void CopyFromPacket(bess::PacketBatch *batch, const struct Attr *attr,
                           mt_offset_t mt_off)

The pkt_off and mt_off are used. Tracing up, the mt_off is offset in the attribute list of that attribute. The line 
int pkt_off = attr->offset;
is also getting the offset of the attribute. So not sure how the offsets calculated.

I am trying to set an attribute 16 bytes long and trying to read in my module. In the next module I see the packet pointer in gdb, I don't see the data I requested to add from SetMetadata at the offset (192) - why I am digging into this. So I would like to understand what needs to pass in offset in attribute struct (of say SetMetadata) and how is mt_off determined? I would need pointer to head to so I can go 192 to the the first attribute in metadata.

**Comment by melvinw** [comment]
Packet metadata offsets are computed when you issue the `ResumeAll()` bessctl command. The entry point to that computation lives [here](https://github.com/NetSys/bess/blob/master/core/metadata.cc#L445). If you start bessd with `-v 1` the logs will show some debugging output from the computation.

As far as reading your metadata attribute goes, there are several [helpers in module.h](https://github.com/NetSys/bess/blob/master/core/module.h#L679) you can use to translate metadata indexes/offsets into meaningful values. You can see an example of their usage in the [IPEncap module](https://github.com/NetSys/bess/blob/master/core/modules/ip_encap.cc) (see the calls to `AddMetadataAttr()` and `get_attr()` in particular).

**Comment by muraliran** [comment]
Thank you. I reviewed the code for write and read attributes and calculate offsets, It was helpful. My requirement is to read attribute in a downstream module. It does not work with this approach either.

I am using set attribute using this

em1:0 -> SetMetadata(attrs=[{'name':'port_id','size': 16,\
            'value_bin': b'\x74\xd4\x36\x79\x7c\x00\x49\x27\x91\xdd\x77\x78\x3e\x65\x2f\x63'}]) ->...

The Init of my next module (sm1  implemented by me) has
int attr_id = AddMetadataAttr("Port_ID", 16,
                            bess::metadata::Attribute::AccessMode::kRead);
That returns 0.

However, the scope is not set right or not sure how that is handled. Probably I am missing that somewhere. How do we use scope?

Attribute does not propagate. When I read using get_attr<>, it seg faults.

From Logs:
I1013 14:10:54.101009  5689 bessctl.cc:534] *** Resuming ***
I1013 14:10:54.101034  5689 metadata.cc:404] scope component for 16-byte attr port_id at offset -1: {
I1013 14:10:54.101039  5689 metadata.cc:409] setattr0
I1013 14:10:54.101044  5689 metadata.cc:412] }
I1013 14:10:54.101048  5689 metadata.cc:404] scope component for 16-byte attr port_id at offset -1: {
I1013 14:10:54.101053  5689 metadata.cc:409] setattr1
I1013 14:10:54.101058  5689 metadata.cc:412] }
....
I1013 14:10:54.101200  5689 metadata.cc:423] Module sm1 part of the following scope components:
W1013 14:10:54.101207  5689 metadata.cc:77] Metadata attr Port_ID/16 of module sm1 has no upstream module that sets the value!

**Comment by muraliran** [comment]
I corrected the character case (Port_ID) but still does not recognize the upstream data set

I1013 14:50:39.313100  5840 metadata.cc:404] scope component for 16-byte attr Port_ID at offset -2: {
I1013 14:50:39.313235  5840 metadata.cc:409] sm1
I1013 14:50:39.313262  5840 metadata.cc:409] pML1_in
I1013 14:50:39.313356  5840 metadata.cc:409] setattr0
I1013 14:50:39.313380  5840 metadata.cc:409] setattr1
I1013 14:50:39.313400  5840 metadata.cc:409] pML2_in
I1013 14:50:39.313421  5840 metadata.cc:412] }
...
..
W1013 14:50:39.314013  5840 metadata.cc:77] Metadata attr Port_ID/16 of module sm1 has no upstream module that sets the value!

So looks like Scope is not the issue.

**Comment by muraliran** [comment]
On a second look, it does look like scoping issue. The attribute when set is indicating the scope. But the module initialized are not showing same scope.  

How do we set scope for a module? 
Or is it implied/set according to the pipeline?

When flow defined in .bess file, I presumed the pipeline would automatically get attached to the modules involved. All modules in the logs show they aren't associated with same scope as defined for attribute. Am I reading it right?

`I1013 14:54:25.189901  5830 metadata.cc:404] scope component for 16-byte attr Port_ID at offset -2: {
I1013 14:54:25.189906  5830 metadata.cc:409] sm1
I1013 14:54:25.189908  5830 metadata.cc:409] pML1_in
I1013 14:54:25.189910  5830 metadata.cc:409] setattr0
I1013 14:54:25.189913  5830 metadata.cc:409] setattr1
I1013 14:54:25.189914  5830 metadata.cc:409] pML2_in
I1013 14:54:25.189934  5830 metadata.cc:412] }
I1013 14:54:25.189937  5830 metadata.cc:423] Module em1 part of the following scope components:
I1013 14:54:25.189940  5830 metadata.cc:423] Module pML1_in part of the following scope components:
I1013 14:54:25.189942  5830 metadata.cc:423] Module pML1_out part of the following scope components:
I1013 14:54:25.189944  5830 metadata.cc:423] Module pML2_in part of the following scope components:
I1013 14:54:25.189946  5830 metadata.cc:423] Module pML2_out part of the following scope components:
I1013 14:54:25.189951  5830 metadata.cc:423] Module rewrite0 part of the following scope components:
I1013 14:54:25.189954  5830 metadata.cc:423] Module s part of the following scope components:
I1013 14:54:25.189955  5830 metadata.cc:423] Module setattr0 part of the following scope components:
I1013 14:54:25.189957  5830 metadata.cc:423] Module setattr1 part of the following scope components:
I1013 14:54:25.189960  5830 metadata.cc:423] Module sink0 part of the following scope components:
I1013 14:54:25.189963  5830 metadata.cc:423] Module sm1 part of the following scope components:
W1013 14:54:25.189966  5830 metadata.cc:77] Metadata attr Port_ID/16 of module sm1 has no upstream module that sets the value!
`

**Comment by melvinw** [comment]
I suspect what's happening here is that there is a path from some other module to `sm1` that has no modules that write to the `Port_ID`. When this happens, the offset computation code will not assign an offset to your `Port_ID`, because it cannot guarantee that you won't be reading/writing to another metadata attribute.

To confirm, can you run `show pipeline --as dot | dot -Tsvg -o pipeline.svg` and upload a copy of `pipeline.svg`?

**Comment by muraliran** [comment]
![pipeline](https://user-images.githubusercontent.com/11051880/31570633-c8f982e6-b03a-11e7-949b-53429c8d3031.png)

Actually the steps I did are pretty similar to ip_encap.cc

In Init I do this:
int attr_id = AddMetadataAttr("Port_ID", 16,
                            bess::metadata::Attribute::AccessMode::kRead);
Checked the return value is 0.

In ProcessBatch I do the get_attr
     get_attr<uint8_t*>(this, 0, pkt)
because the attr_id is 0. It is 16 bytes. Based on the get_attr<> code, it is returning the pointer in pkt at the offset so think the casting should work. But it faults there - so it was a bit tricky to get the screen shot of pipeline pausing with gdb :) 
The command this->get_offset(attr_id) seems to be valid but not sure if it errors due to array cast in template.

The logs are in response to Init AddMetadataAttr logs as 
I1013 17:15:44.738291   867 metadata.cc:423] Module sm1 part of the following scope components:
W1013 17:15:44.738296   867 metadata.cc:77] Metadata attr Port_ID/16 of module sm1 has no upstream module that sets the value!

So I think the initialization itself is erroring unable to recognize the Port_ID set upstream. I don't see any code in test that tests an array of bytes being passed from one module to another so this could be a new case.
I1013 17:15:44.738021   867 bessctl.cc:534] *** Resuming ***
I1013 17:15:44.738118   867 metadata.cc:404] scope component for 16-byte attr Port_ID at offset -2: {
I1013 17:15:44.738133   867 metadata.cc:409] pML1_in
I1013 17:15:44.738219   867 metadata.cc:409] pML2_in
I1013 17:15:44.738227   867 metadata.cc:409] sm1
I1013 17:15:44.738231   867 metadata.cc:409] setattr0
I1013 17:15:44.738235   867 metadata.cc:409] setattr1
I1013 17:15:44.738240   867 metadata.cc:412] }

I see the notes in a closed case 2 days back on Metadata theory (slides) - which looks fine to me. But not sure how the sub-graph or the scope component works dynamically.

**Comment by muraliran** [comment]
Tried changing to 1 byte int, same results. Not sure how to create the scope component to have continuity

I1013 19:39:21.667400 22144 bessctl.cc:695] Checking scheduling constraints
I1013 19:39:21.671852 22144 bessctl.cc:534] *** Resuming ***
I1013 19:39:21.671890 22144 metadata.cc:404] scope component for 1-byte attr PortNo at offset -2: {
I1013 19:39:21.671895 22144 metadata.cc:409] setattr0
I1013 19:39:21.671900 22144 metadata.cc:409] setattr1
I1013 19:39:21.671905 22144 metadata.cc:409] pML2_in
I1013 19:39:21.671910 22144 metadata.cc:409] pML1_in
I1013 19:39:21.671913 22144 metadata.cc:409] sm1
I1013 19:39:21.671917 22144 metadata.cc:412] }
I1013 19:39:21.671921 22144 metadata.cc:423] Module em1 part of the following scope components:
I1013 19:39:21.671926 22144 metadata.cc:423] Module pML1_in part of the following scope components:
I1013 19:39:21.671931 22144 metadata.cc:423] Module pML1_out part of the following scope components:
I1013 19:39:21.671936 22144 metadata.cc:423] Module pML2_in part of the following scope components:
I1013 19:39:21.671941 22144 metadata.cc:423] Module pML2_out part of the following scope components:
I1013 19:39:21.671944 22144 metadata.cc:423] Module rewrite0 part of the following scope components:
I1013 19:39:21.671948 22144 metadata.cc:423] Module s part of the following scope components:
I1013 19:39:21.671953 22144 metadata.cc:423] Module setattr0 part of the following scope components:
I1013 19:39:21.671957 22144 metadata.cc:423] Module setattr1 part of the following scope components:
I1013 19:39:21.671962 22144 metadata.cc:423] Module sink0 part of the following scope components:
I1013 19:39:21.671967 22144 metadata.cc:423] Module sm1 part of the following scope components:
W1013 19:39:21.672039 22144 metadata.cc:77] Metadata attr PortNo/1 of module sm1 has no upstream module that sets the value!


---

Flow in the .bess

em1.add(fields=[{'value_int':17}, {'value_bin':aton('10.20.21.11')}], gate=0)
em1.add(fields=[{'value_int':17}, {'value_bin':aton('10.20.21.12')}], gate=1)
....
.. (few entries like this of different proto).
..
em1:0 -> SetMetadata(attrs=[{'name':'PortNo','size':1,'value_int': 0x01}]) -> 0:sm1
em1:1 -> SetMetadata(attrs=[{'name':'PortNo','size':1,'value_int': 0x02}]) -> 0:sm1
 
The code:

Init:
    AddMetadataAttr("PortNo", 1, bess::metadata::Attribute::AccessMode::kRead);

ProcessBatch
    const uint8_t port_no = get_attr<uint8_t>(this, 0, pkt);
    out_gates[i] = port_no;

port_no is always 0

**Comment by melvinw** [comment]
Ah, the screenshot of `show pipeline` confirms my suspicion! You need to add a module that sets `PortNo` between `sm1` and both of `pML1_in` and `pML2_in`.

The offset computation code is currently assigning `PortNo` an offset of `-2`, which roughly means there isn't a writer of `PortNo` on every path to the module that reads it (`sm1` in this case). The metadata system does this because to make sure readers of metadata attributes don't read garbage. This is an unfortunate and slightly frustrating property of the current metadata system. While `SmartSwitch::ProcessBatch()` might be smart enough to only read from `PortNo` for packets from input gate 0, there is currently no way to communicate that to the rest of BESS.

**Comment by muraliran** [comment]
It does not make sense. The pML1 and pML2 are vports - I can't add same attributes there. However, I do expect the attributes to stay alive so I can make associations at pML1_in at the switch. We need to enhance the TraverseUpstream and downstream to work more intelligently. I was just going through those methods - it isn't invoked from many places why the scope is not getting built well. I was looking how to do traverse so when an attribute is added before the pipeline created also traverses - which seems to be the issue at this time for me here

**Comment by melvinw** [comment]
Yes, you're right. You can't add metadata attributes to `PortInc`. You can, however, add another `SetMetadata` module between `pML1_in` and `sm1` that sets `PortNo` to some meaningful value (same with `pML2_in`).  Based on what you've sent so far, I'm guessing you want to replace the following:

```
pML1_inc:0 -> 1:sm1
pML2_inc:0 -> 2:sm1
```

with:

```
pML1_inc:0 -> SetMetadata(attrs=[{'name':'PortNo','size':1,'value_int': 0x01}]) -> 1:sm1
pML2_inc:0 -> SetMetadata(attrs=[{'name':'PortNo','size':1,'value_int': 0x02}]) -> 2:sm1
```

If you've got a change to the metadata code in mind, please do write it up and submit a pull request! Also, have a look at [these slides](https://github.com/NetSys/bess/files/1376696/BESS_metadata.pptx) for some background on the current design.

**Comment by muraliran** [comment]
Thanks Melvin. Yes, eventually I will have my own modules and not be using set_metadata. The idea is to maintain a unique id (uuid) for each port (not portNo) (like in openstack) so the flow definitions are much easier. My thinking is at [ResumeAll](https://github.com/NetSys/bess/blob/master/core/metadata.cc#L445) you mentioned in first comment above, needs also do a traverse as pipelines are established by then. For now that looks like a logical point.. Let me check at my end first if it will work.

I also see your wish list on fully contained .so (linking) - I would like that too. While integrating uuid libs, I had to do static link for all modules. I will comment on that issue later.

**Comment by muraliran** [comment]
![pipeline1](https://user-images.githubusercontent.com/11051880/31588739-29a8c108-b1ab-11e7-8cb0-ce3ce59b91a8.png)

Just some updates...

I had to add set_metadata everywhere - all outgoing & incoming paths to SM1 - to be able to get the metadata in sm1. This implementation is too restrictive, really useless. The metadata should be available until a read exists in any one path. I tried adding the TraverseUpstream in Pipeline::IdentifyScopeComponent thinking maybe that will resolve but did not work.

So basically the Traverse upstream and downstream logic needs to change. It will take some time for me to figure how to get it working.

**Comment by muraliran** [comment]
You can see in the picture the bytes are travelling to pML1 and pML2 - indication that the PortNo was read at sm1.

**Comment by melvinw** [comment]
Ah, sorry-- I may have closed this issue prematurely.

To achieve the behavior you want you will need to change quite a bit more than just `TraverseUpstream()` and `TraverseDownstream()`. The fix there is actually trivial-- just delete [this if-statement](https://github.com/NetSys/bess/blob/master/core/metadata.cc#L220). However, without that if-statement you have zero guarantees that readers of metadata attributes won't be reading garbage. Providing said guarantees must be done carefully, as I think they will impose non-negligible per-packet overhead.

One way to do this would be to change the API to allow an optional default value when registering an attribute. This should allow only readers (at the gate level, hopefully) of such attributes to incur the overheads of making sure attributes are properly initialized.

Another approach would be to just assign metadata offsets globally on a first-come-first-serve basis. This is the most straightforward approach, but it limits efficient sharing of the limited metadata buffer-- hence the current design.

**Comment by muraliran** [comment]
The fact that an attribute read was registered should be enough reason to sustain the longevity of the attribute. I still don't comprehend fully how the pipelines are determined (in the code) and defining scope need to match that. In ovs pipeline is from ingress to egress relative to the switch but in BESS that is not really well defined. I am not against the concepts in the slide-deck but the way it is currently implemented is not reflective of the concepts. The scope should be defined in the pipeline from the time an attribute created until it is not used any longer. There should be no need to define attribute in every data paths from start to end - then there is no sub-graph. A traverse up and/or down on resume-all should be able to resolve/determine the scope for the attribute. A scope needs to be subset of pipeline, if I understand pipelines correctly here.
As for garbage read, at this time I get garbage read of non-scalar attribute (such as 16 byte array) as we process a batch. I will probably look into that first to get an appropriate helper and then into the scope resolution later this week. Or if yourself or anyone else has better ideas, please share or implement.

**Comment by melvinw** [comment]
I don't understand the disconnect here. What modules in your [first screenshot](https://user-images.githubusercontent.com/11051880/31570633-c8f982e6-b03a-11e7-949b-53429c8d3031.png) do you expect to be a part of the scope of `PortNo`?

**Comment by muraliran** [comment]
![metadata_working](https://user-images.githubusercontent.com/11051880/31630928-ffc0c6de-b26c-11e7-8129-8b7a5dafe1b1.png)
Melvin, this pic should show the difference. I am expecting the first pipeline to work. I don't have to use beyond a module for it to be available in an earlier module.

From same pic let us say bubbles 1 through 4.

Set in 1 -- read in 4   - should be in scope (should work)
Set in 2 -- read in 3  - should work
Set in 1 -- no access in 2  -- modify in 3 -- read in 4 should work even if 5, 6, 7 are not using them.

Currently the scope definition is one thing - but more importantly I am not sure what is a pipeline. as implemented now 

I would think, a graph pat from port_inc to port_out should be a pipeline because that is a scheduled item.

If there is a path consider the p0 to be dpdk port, vc1, vc2, vc3 are vports to containers on same or different hosts:
Then:
  p0 -> sw1 -> ip_lookup -> vc1_port_out   -- is one pipeline

Notwithstanding the fact that I could also have:
     vc1_port_inc - > sw1 -> ip_lookup -> p0_port_out -> host2_p0_int -> l2_forward -> vc2_port_out

     vc2_port_inc   ->   sw2  -> ip_lookup -> vxlan_encap -> vxlan_decap -> vc3_port_out

Each of which should be a pipeline of its own. But that is not how it is drawn currently in "show pipeline". It includes paths of vport in & out inter-twined together as in my 2nd screen shot.

So prior to fixing the scope, I think we need to define the pipeline properly. Please let me know if any of my observation is incorrect or if the pipeline definition as it is now makes sense.

**Comment by muraliran** [comment]
I think I am chewing more than needed :) Let us first resolve on one host. Just think vc1, vc2 are on same host and we define 2 distinct pipelines from 
p0 -> sw1 -> vc1_port_inc  and
vc1_port_inc -> sw1 ->  vc2_port_out as 2 schedulable pipelines - that may pass through same module in the first pipeline. Currently it is shown all the paths as one pipeline.

**Comment by melvinw** [comment]
The current system does indeed support the first pipeline you posted.

Consider the script below. The `MetadataTest` module configures itself as either a reader or writer of an attribute for testing purposes.

```
src::Source()
writer::MetadataTest(write={'a':2, 'b': 2})
reader::MetadataTest(read={'a':2, 'b': 2})
sink::Sink()

src -> writer -> reader -> sink
```

When executed both attributes are assigned valid offsets.

```
localhost:10514 $ show module
  reader::MetadataTest()
    Per-packet metadata fields:
              a: read   2 bytes at offset 0
              b: read   2 bytes at offset 2
    Input gates:
          0: batches 0                packets 0                writer:0 ->
    Output gates:
          0: batches 6202208          packets 198470656        -> 0:sink
  sink::Sink()
    Input gates:
          0: batches 0                packets 0                reader:0 ->
  src::Source()
    Output gates:
          0: batches 6205198          packets 198566336        -> 0:writer
  writer::MetadataTest()
    Per-packet metadata fields:
              a: write  2 bytes at offset 0
              b: write  2 bytes at offset 2
    Input gates:
          0: batches 0                packets 0                src:0 ->
    Output gates:
          0: batches 6206357          packets 198603424        -> 0:reader
```

```
localhost:10514 $ show pipeline
+--------+                     +--------------+                     +--------------+                     +------+
|  src   |  :0 1009289856 0:   |    writer    |  :0 1009327616 0:   |    reader    |  :0 1009203488 0:   | sink |
| Source | ------------------> | MetadataTest | ------------------> | MetadataTest | ------------------> | Sink |
+--------+                     +--------------+                     +--------------+                     +------+
```

You cannot safely consider individual paths through the graph as the entire scope of an attribute-- as there may be more than one path to a given reader (which I think is the behavior you're describing now). If you modify the above script to the following:

```
src::Source()
writer::MetadataTest(write={'a':2, 'b': 2})
reader::MetadataTest(read={'a':2, 'b': 2})
sink::Sink()
src1::Source()

src -> writer -> reader -> sink
src2 -> reader
```

Then the metadata system, correctly, considers both attributes invalid because it has no idea what to do with packets going from `src2` to `reader`.

```
localhost:10514 $ show module
  reader::MetadataTest()
    Per-packet metadata fields:
              a: read   2 bytes (no upstream writer)
              b: read   2 bytes (no upstream writer)
    Input gates:
          0: batches 0                packets 0                writer:0 ->, src2:0 ->
    Output gates:
          0: batches 37976589         packets 1215250848       -> 0:sink
  sink::Sink()
    Input gates:
          0: batches 0                packets 0                reader:0 ->
  src::Source()
    Output gates:
          0: batches 18989695         packets 607670240        -> 0:writer
  src2::Source()
    Output gates:
          0: batches 18990256         packets 607688192        -> 0:reader
  writer::MetadataTest()
    Per-packet metadata fields:
              a: write  2 bytes (no downstream reader)
              b: write  2 bytes (no downstream reader)
    Input gates:
          0: batches 0                packets 0                src:0 ->
    Output gates:
          0: batches 18990801         packets 607705632        -> 0:reader
localhost:10514 $ show pipeline
+--------+                    +------------------+                     +------+
|  src2  |  :0 863311616 0:   |      reader      |  :0 1726511808 0:   | sink |
| Source | -----------------> |   MetadataTest   | ------------------> | Sink |
+--------+                    +------------------+                     +------+
                                ^
                                | :0 863327840 0:
                                |
+--------+                    +------------------+
|  src   |  :0 863294560 0:   |      writer      |
| Source | -----------------> |   MetadataTest   |
+--------+                    +------------------+
```

Anyway... I plan to incorporate support for the second script in the metadata refactoring. For now you'll just need to add `SetMetadata` modules everywhere. It's annoying-- I know :)

**Comment by chris3torek** [comment]
Let me also add here that this is somewhat like compilers complaining about this code:

    void f(int arg) {
        int i, j;
        if (arg == 1) {
            i = j = 0;
        }
        // ... code that does not use or set i and j
        if (arg > 0) {
            // code that does use i and/or j
            // some compilers complain
        }
    }

Here the problem space is a bit different, but the problem itself is very similar: the compiler (or metadata-assigner) cannot tell that (or maybe I should say _if_) `arg` is always exactly 0 or 1, and therefore `arg == 1` and `arg > 0` both "mean the same thing".  The compiler _can_, however, tell that if `arg == 2` at entry, the second section of code uses `i` and `j` when they were never set to any initial value.  In this case, the behavior of the code is not predictable, so the compiler might just optimize the variables out entirely (whether this is allowed depends on the language: if the above is C or C++ it's not actually allowed since the code behaves properly if `arg` is always 0 or 1).

In terms of the metadata "language", it's (at least currently, based on behavior) considered undefined.  Changing the source-equivalent to just always assign 0, though, makes the metadata "compiler" happy with it: `i` and `j` will get valid offsets assigned.

**Comment by muraliran** [comment]
I sort of defer from Melvin's comment. From the bubble graph I attached - the first one should have been how my use case would be if pipeline was defined properly - lets get to that later. The egress from pML1 to SM1 should be in a different pipeline - logically.

To illustrate better, let me take one of your examples.

![complicated_exaple](https://user-images.githubusercontent.com/11051880/31638352-fb13cf76-b287-11e7-9fbf-c75fb7039518.png)


We need to think of how packets are handled in a network, so the scenario I was stating (my 2nd screen shot) is not out of the world. From the pic attached, here is what needs to happen. I am just thinking logically here, not as an academician on how compilers or some graph theory must work.

At this time I have somewhat similar to the graph showing blocks A1, A2... which I want to be in a scope based on attributes write/read. Here is how I think it needs to work to be useful. Module A4 should be able to read B, E and D without getting de-scoped.

B is carried unmodified from box before A1 and de-scoped at A5. However The scope of D is different. We can assume all Ai boxes in one scope or in 2 scopes. I would think it should be in 2 scopes. The attr D & E are a bit tricky because when packets come from A1 to A2 to A4 the values should be reading NULL. The fact that there is a possibility of having D set from A7, the Traverse Upstream must set the attribute field at A2 (meaning offset available). As a general notion, metadata area can be zero pre-filled at start of pipeline ( pipeline :) ) , so any read will yield zero, not garbage. But when offset is reused, it is not guaranteed which is what you indicate as a problem.  I would leave such things to the programmer because metadata is not like a user mode scratch pad where anything can be written or a well defined rule needs to be forced. Attributes have to be something that is fairly well defined during the data plane design, so any convention on how to handle reasonably should be fine. Rather than proactively de-scope, I would risk having it around and leave it to the user to manage - like memory mgmt in C programming. If a traverse sees no use of attribute from a module (not including the last used), the it can de-scoped from there. We can set a config param to do proactive de-scope if that functionality is important. For something like in A4, in my code I would read D only if C & D are set or E & D are set. It is up to the programmer to define the scope and infer data. Will that be reasonable?

In my case I will be using some identifiers that will be available throughout the framework (very similar to ovs but more flexible), so don't anticipate very active attr set/read/unset etc like in this complex example. I can't foresee such "busy" activity as in your slide number 15 in a typical network data plane, but maybe you folks have such scenario, I don't know. Most items in slide number 3 are standard attrs that will be constantly used in the data plane. For any application (vnf), one may use a smaller set of all options you folks listed in slide number 3 and can easily be scoped within a certain graph of modules. In the worst case insert a module that clears the fields like SetMetadata.

My other issue was with the pipeline definition, maybe will start a new thread on that once I dig in more on how it is defined now.

**Comment by chris3torek** [comment]
Just to back off to a very high level view, if you're saying that the current setup is very hard to program (and difficult to reason about), I'd have to agree!  Carrying the language analogy perhaps too far, it's like the bad old Fortran days when you could declare multiple entry points to a function, and had to be very careful to set everything up along every possible entry point.

How to deal with this better, in terms of meta-programming, is a bit of an open question.  One possibility is to declare that any read-before-write _automatically_ inserts a pre-write of zero, perhaps along with a warning:

                              +--------+
    <source without set> ---> | module |
    <source with set A=0> --> | read A |
                              +--------+

would become

                                       +--------+
    <source without set> --> <A=0> --> | module |
    <source with set A=0> -----------> | read A |
                                       +--------+

Anywhere any of these warnings occurred (if there are warnings) show you where you perhaps forgot to set the metadata that the downstream module will read.

But this does not help, and maybe even hurts a bit, in more complex cases:

                                       +--------+     +--------------+
    <source without set> --> <A=0> --> | module |     | read A       |
    <source with set A=0> -----------> | read A | --> | if A, read B |
                                       +--------+     +--------------+
                                                              ^
    <source with set A=1, B=...>------------------------------+

Here, because there is a read(B) op, all "upstream" modules would be forced to write B and the system would insert a pipeline stage that does that (the logical place to do that is immediately before the module that uses it, hence between the two larger boxes in this diagram).  That's actually a waste of time since the read in the second large box module is conditional—but we do not and cannot know that, given the current setup.

Another option is simply to warn: if there's a read "downstream" of some path that has no write, emit the warning and pretend that the variable was defined, but don't actually initialize it at all.

**Comment by muraliran** [comment]
The reason it is looking complicated is because it is made complicated by defining pipeline encompassing everything in the graph. I am not thinking it is complicated, but it does not solve problems in real life. I am a pragmatic engineer :-)

I am working on a strategy to define pipelines. A graph path should be well defined and (like namespace) assigned to a pipeline. There won't be multiple paths you have to deal with in that case. An attribute is defined on a pipeline so it will be scoped - like:
p1::attr1
p2::attr2 .. etc

A module should be able to support more than one pipelines. I have started to code along these lines and let me see how it goes in a couple of days. The module class has a vector of pipelines. When a graph is defined, it must be qualified by a pipeline. Rest of the logic as it is now for attribute might just work or with minor changes. I need to see if this will create any issue. Most likely not because at this time everything is stuffed into default pipeline. And get_pipeline() is not used anywhere.
